llvm/lib/SYCLLowerIR/SYCLVirtualFunctionsAnalysis.cpp

```
@@ -110,7 +112,8 @@ void collectVTablesThatUseFunction(
       //    ptr addrspace(4) addrspacecast (ptr @foo to ptr addrspace(4)), ...]
       collectVTablesThatUseFunction(U, VTables);
     } else {
-      llvm_unreachable("Unhandled type of user");
```

According to my understanding, `llvm_unreachable` is put here to catch cases which were missed during development. Ignoring "unhandled" cases might lead to bugs which are hard to analyze. The crash here is easy to analyze.

@Maetveis, do you have test case leading to running into `llvm_unreachable`? If so, please, add a regression test and one more if to handle it. 

If would prefer to have `llvm_unreachable("Unhandled type of user");` for unhandled users instead of silently ignoring them.

---

See the attached test, it was reduced from basically
```cpp
struct C {
/* sycl_indirectly_callable */
void foo(){)
};

SYCL_EXTERNAL void direct_caller {
  C c;
  c.foo();
}
```

I can add an `isa<Instruction>` branch if thats preferred.

---

I think it's preferred and will help us detecting incompatible LLVM community changes (e.g. adding more types of Users we need to handle).

Alternatively, we can consider replacing `llvm_unreachable` with an `assert`, but I'd like code owners/authors to comment on that first.

---

The code is trying to look through constant expressions to find global variables that use the function pointers in their initializers. Personally, I don't really expect LLVM to start allowing non constants in global variable initializers). If other `User`s appear they are either going to be subclasses of `Constant` or probably won't appear in global initializers.
BTW looking at it again even the current code is redundant, `ConstantExpr` **is a** subclass of `Constant`.

IMO having an allowlist here is unlikely to catch real bugs, but can easily be a source of churn or hidden crashes.

---

> ```c++
> struct C {
>   /* sycl_indirectly_callable */
>   void foo(){}
> };
> 
> SYCL_EXTERNAL void direct_caller() {
>   C c;
>   c.foo();
> }
> ```
>
> Should this be added as test too?

What is the form of `sycl_indirectly_callable`? Right know there is only a single (more or less properly) documented way of using indirect calls in SYCL - that is through virtual functions and the corresponding extension (#10540).

---

Thanks for providing the reproducer. That is indeed a valid use case (i.e. a non-virtual call to a virtual function) and we shouldn't crash.

Reading documentation on [`llvm_unreachable`](https://llvm.org/doxygen/llvm_2Support_2ErrorHandling_8h.html#ace243f5c25697a1107cce46626b3dc94) I see that it the behavior of reaching it is effectively undefined which isn't good for our release builds.

From that point of view, its the right thing to move away from it. However, I still think that its a valuable information for us as developers to know about potentially missing cases in our virtual functions analysis. From that point of view, using debug macro isn't the ideal approach, because it requires an opt-in.

I would prefer to have this fixed by transforming the unreachable into an `assert`:
```c++
  assert(isa<CallInst>(U) && "Unexpected user of a virtual function");
```

That way anything unexpected will be ignored in release build, but it will be immediately highlighted for us.

---

> I would prefer to have this fixed by transforming the unreachable into an `assert`:
> ```cpp
>   assert(isa<CallInst>(U) && "Unexpected user of a virtual function");
> ```

This was also my original approach before posting the PR, but there's a problem I realized quickly: `@function` is perfectly valid pointer constant in the IR, and can be used anywhere a pointer can be, calls are just the most common example. The following is all well-formed (though maybe UB) IR:

```llvm
declare i32 @square(i32)

define void @test() {
    %1 = ptrtoint ptr @square to i64
    %2 = call i32 @square(i32 0)
    ; These are probably UB, but should not crash the compiler!
    %3 = load i32, ptr @square
    %4 = addrspacecast ptr @square to ptr addrspace(1) ;
    ; invoke could be common too, but it has complex requirements, can't show it here
    ret void
}
```

So if we don't want to take on maintenance of a list of instructions that can take pointer operands then the assert would need to be:

```cpp
   assert(isa<Instruction>(U) && "Unexpected user of a virtual function");
```

I can add that if you're okay with it, though I find it of dubious value.